### ISSUE TITLE
fatal error: concurrent map iteration and map write

### ISSUE THREAD

**Issue by Blaxon** [qiniu/goc#324]
My goc server serves 300+ services at the same time, however it gets killed when handling multi requests
return fatal error.
Could you help check the issue? Thanks!

goc version: v1.4.4
```
11253536-[GIN] 2023/02/14 - 11:59:45 | 200 |     379.535µs |  10.129.106.106 | POST     "/v1/cover/remove"
11253635-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://127.0.0.1:30963] not found" file="[cover/server.go:422]"
11253760-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://172.17.0.1:30963] not found" file="[cover/server.go:422]"
11253886-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://169.254.1.1:30963] not found" file="[cover/server.go:422]"
11254013-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://10.105.190.124:30963] not found" file="[cover/server.go:422]"
11254143-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.0.1:30963] not found" file="[cover/server.go:422]"
11254270-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.238.222:30963] not found" file="[cover/server.go:422]"
11254401-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.0.10:30963] not found" file="[cover/server.go:422]"
11254529-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.104.196:30963] not found" file="[cover/server.go:422]"
11254660-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.26.8:30963] not found" file="[cover/server.go:422]"
11254788-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.176.60:30963] not found" file="[cover/server.go:422]"
11254918-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.68.156:30963] not found" file="[cover/server.go:422]"
11255048-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.96.113:30963] not found" file="[cover/server.go:422]"
11255178-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.62.249:30963] not found" file="[cover/server.go:422]"
11255308-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.53.222:30963] not found" file="[cover/server.go:422]"
11255438-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.111.144:30963] not found" file="[cover/server.go:422]"
11255569-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.159.166:30963] not found" file="[cover/server.go:422]"
11255700-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.38.38:30963] not found" file="[cover/server.go:422]"
11255829-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.43.23:30963] not found" file="[cover/server.go:422]"
11255958-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://127.0.0.1:22899] not found" file="[cover/server.go:422]"
11256083-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://172.17.0.1:22899] not found" file="[cover/server.go:422]"
11256209-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.43.23:22899] not found" file="[cover/server.go:422]"
11256338-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.104.196:22899] not found" file="[cover/server.go:422]"
11256469-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.62.249:22899] not found" file="[cover/server.go:422]"
11256599-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.26.8:22899] not found" file="[cover/server.go:422]"
11256727-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.68.156:22899] not found" file="[cover/server.go:422]"
11256857-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.38.38:22899] not found" file="[cover/server.go:422]"
11256986-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.176.60:22899] not found" file="[cover/server.go:422]"
11257116-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.159.166:22899] not found" file="[cover/server.go:422]"
11257247-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.0.10:22899] not found" file="[cover/server.go:422]"
11257375-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.53.222:22899] not found" file="[cover/server.go:422]"
11257505-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.96.113:22899] not found" file="[cover/server.go:422]"
11257635-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.111.144:22899] not found" file="[cover/server.go:422]"
11257766-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.238.222:22899] not found" file="[cover/server.go:422]"
11257897-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://192.168.0.1:22899] not found" file="[cover/server.go:422]"
11258024-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://169.254.1.1:22899] not found" file="[cover/server.go:422]"
11258151-time="2023-02-14T11:59:46+08:00" level=warning msg="address [http://10.105.150.27:22899] not found" file="[cover/server.go:422]"
11258280:fatal error: concurrent map iteration and map write
11258332-
11258333-goroutine 65059 [running]:
11258360-github.com/qiniu/goc/pkg/cover.(*fileStore).Set(0xc0003447c0, 0xc000146940?)
11258437-	/home/runner/work/goc/goc/pkg/cover/store.go:186 +0x1d3
11258494-github.com/qiniu/goc/pkg/cover.(*fileStore).Remove(0xc0003447c0, {0xc000146940?, 0xc0001e2380?})
11258591-	/home/runner/work/goc/goc/pkg/cover/store.go:113 +0x5e
11258647-github.com/qiniu/goc/pkg/cover.(*server).removeServices(0xc00039c630, 0xc0001d2300)
11258731-	/home/runner/work/goc/goc/pkg/cover/server.go:353 +0x342
11258789-github.com/gin-gonic/gin.(*Context).Next(...)
11258835-	/home/runner/go/pkg/mod/github.com/gin-gonic/gin@v1.7.2/context.go:165
11258907-github.com/gin-gonic/gin.CustomRecoveryWithWriter.func1(0xc0001d2300)
11258977-	/home/runner/go/pkg/mod/github.com/gin-gonic/gin@v1.7.2/recovery.go:99 +0x82
11259055-github.com/gin-gonic/gin.(*Context).Next(...)
11259101-	/home/runner/go/pkg/mod/github.com/gin-gonic/gin@v1.7.2/context.go:165
11259173-github.com/gin-gonic/gin.LoggerWithConfig.func1(0xc0001d2300)
11259235-	/home/runner/go/pkg/mod/github.com/gin-gonic/gin@v1.7.2/logger.go:241 +0xe7
11259312-github.com/gin-gonic/gin.(*Context).Next(...)
11259358-	/home/runner/go/pkg/mod/github.com/gin-gonic/gin@v1.7.2/context.go:165
11259430-github.com/gin-gonic/gin.(*Engine).handleHTTPRequest(0xc000492680, 0xc0001d2300)
11259511-	/home/runner/go/pkg/mod/github.com/gin-gonic/gin@v1.7.2/gin.go:489 +0x63e
11259586-github.com/gin-gonic/gin.(*Engine).ServeHTTP(0xc000492680, {0xccf790?, 0xc000118460}, 0xc0001d2a00)
11259686-	/home/runner/go/pkg/mod/github.com/gin-gonic/gin@v1.7.2/gin.go:445 +0x1c5
11259761-net/http.serverHandler.ServeHTTP({0xccd2a0?}, {0xccf790, 0xc000118460}, 0xc0001d2a00)
11259847-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/http/server.go:2947 +0x30c
11259918-net/http.(*conn).serve(0xc0003a26e0, {0xcd0240, 0xc00039cc00})
11259981-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/http/server.go:1991 +0x607
11260052-created by net/http.(*Server).Serve
11260088-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/http/server.go:3102 +0x4db
11260159-
11260160-goroutine 1 [IO wait]:
11260183-internal/poll.runtime_pollWait(0x7f078ce2fec8, 0x72)
11260236-	/opt/hostedtoolcache/go/1.19.0/x64/src/runtime/netpoll.go:305 +0x89
11260305-internal/poll.(*pollDesc).wait(0xc00011c280?, 0x6?, 0x0)
11260362-	/opt/hostedtoolcache/go/1.19.0/x64/src/internal/poll/fd_poll_runtime.go:84 +0x32
11260444-internal/poll.(*pollDesc).waitRead(...)
11260484-	/opt/hostedtoolcache/go/1.19.0/x64/src/internal/poll/fd_poll_runtime.go:89
11260560-internal/poll.(*FD).Accept(0xc00011c280)
11260601-	/opt/hostedtoolcache/go/1.19.0/x64/src/internal/poll/fd_unix.go:614 +0x234
11260677-net.(*netFD).accept(0xc00011c280)
11260711-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/fd_unix.go:172 +0x35
11260776-net.(*TCPListener).accept(0xc00011a4f8)
11260816-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/tcpsock_posix.go:142 +0x28
11260887-net.(*TCPListener).Accept(0xc00011a4f8)
11260927-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/tcpsock.go:288 +0x3d
11260992-net/http.(*Server).Serve(0xc0001c6000, {0xccf580, 0xc00011a4f8})
11261057-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/http/server.go:3070 +0x385
11261128-net/http.(*Server).ListenAndServe(0xc0001c6000)
11261176-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/http/server.go:2999 +0x7d
11261246-net/http.ListenAndServe(...)
11261275-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/http/server.go:3255
11261339-github.com/gin-gonic/gin.(*Engine).Run(0xc000492680, {0xc00041dcf0, 0x1, 0x1})
11261418-	/home/runner/go/pkg/mod/github.com/gin-gonic/gin@v1.7.2/gin.go:336 +0x198
11261493-github.com/qiniu/goc/pkg/cover.(*server).Run(0x7ffc59803d46?, {0x7ffc59803d25, 0xc})
11261578-	/home/runner/work/goc/goc/pkg/cover/server.go:76 +0x39c
11261635-github.com/qiniu/goc/cmd.glob..func15(0x113cc40?, {0xba92f4?, 0x3?, 0x3?})
11261710-	/home/runner/work/goc/goc/cmd/server.go:46 +0x8b
11261760-github.com/spf13/cobra.(*Command).execute(0x113cc40, {0xc00039c240, 0x3, 0x3})
11261839-	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846 +0x663
11261916-github.com/spf13/cobra.(*Command).ExecuteC(0x113c700)
11261970-	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x39d
11262047-github.com/spf13/cobra.(*Command).Execute(...)
11262094-	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
11262164-github.com/qiniu/goc/cmd.Execute()
11262199-	/home/runner/work/goc/goc/cmd/root.go:85 +0x25
11262247-main.main()
11262259-	/home/runner/work/goc/goc/goc.go:22 +0x17
11262302-
11262303-goroutine 29798 [IO wait, 36 minutes]:
11262342-internal/poll.runtime_pollWait(0x7f078ce2fb08, 0x72)
11262395-	/opt/hostedtoolcache/go/1.19.0/x64/src/runtime/netpoll.go:305 +0x89
11262464-internal/poll.(*pollDesc).wait(0xc0000a0500?, 0xc0004e6000?, 0x0)
11262530-	/opt/hostedtoolcache/go/1.19.0/x64/src/internal/poll/fd_poll_runtime.go:84 +0x32
11262612-internal/poll.(*pollDesc).waitRead(...)
11262652-	/opt/hostedtoolcache/go/1.19.0/x64/src/internal/poll/fd_poll_runtime.go:89
11262728-internal/poll.(*FD).Read(0xc0000a0500, {0xc0004e6000, 0x1000, 0x1000})
11262799-	/opt/hostedtoolcache/go/1.19.0/x64/src/internal/poll/fd_unix.go:167 +0x25a
11262875-net.(*netFD).Read(0xc0000a0500, {0xc0004e6000?, 0x8?, 0x30?})
11262937-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/fd_posix.go:55 +0x29
11263002-net.(*conn).Read(0xc00001c038, {0xc0004e6000?, 0x203000?, 0x203000?})
11263072-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/net.go:183 +0x45
11263133-net/http.(*connReader).Read(0xc000483aa0, {0xc0004e6000, 0x1000, 0x1000})
11263207-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/http/server.go:786 +0x171
11263277-bufio.(*Reader).fill(0xc000480a20)
11263312-	/opt/hostedtoolcache/go/1.19.0/x64/src/bufio/bufio.go:106 +0xff
11263377-bufio.(*Reader).ReadSlice(0xc000480a20, 0x0?)
11263423-	/opt/hostedtoolcache/go/1.19.0/x64/src/bufio/bufio.go:372 +0x2f
11263488-bufio.(*Reader).ReadLine(0xc000480a20)
11263527-	/opt/hostedtoolcache/go/1.19.0/x64/src/bufio/bufio.go:401 +0x27
11263592-net/textproto.(*Reader).readLineSlice(0xc0037c8de0)
11263644-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/textproto/reader.go:56 +0x99
11263717-net/textproto.(*Reader).ReadLine(...)
11263755-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/textproto/reader.go:37
11263822-net/http.readRequest(0xc00001c038?)
11263858-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/http/request.go:1036 +0x79
11263929-net/http.(*conn).readRequest(0xc000252640, {0xcd0198, 0xc0000469c0})
11263998-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/http/server.go:994 +0x24a
11264068-net/http.(*conn).serve(0xc000252640, {0xcd0240, 0xc00039cc00})
11264131-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/http/server.go:1916 +0x345
11264202-created by net/http.(*Server).Serve
11264238-	/opt/hostedtoolcache/go/1.19.0/x64/src/net/http/server.go:3102 +0x4db
```

**Comment by Blaxon** [comment]
Hi Issue was solve in here: https://github.com/qiniu/goc/pull/329

**Comment by Blaxon** [comment]
The issue was introduced in https://github.com/qiniu/goc/blob/master/pkg/cover/store.go#L107

when we set the memoryStore list back to fileStore. https://github.com/qiniu/goc/blob/master/pkg/cover/store.go#L113
the `list l.memoryStore.GetAll() -> services` passed into `func (l *fileStore) Set(services map[string][]string) error`
in [L175](https://github.com/qiniu/goc/blob/master/pkg/cover/store.go#L175) we sign the services directly to `l.memoryStore.servicesMap`, without copying it.

while in [L186](https://github.com/qiniu/goc/blob/master/pkg/cover/store.go#L186) we are still using this map to do the iteration without locker. (by right should use memoryStore.mu)

The issue will happen when there is another goroutine tries to remove the server from memoryStore. Map will clash.

**Comment by CarlJi** [comment]
fixed by #329
